### DISCUSSION TITLE
Future of Beaming

### DISCUSSION THREAD

**Discussion by TylerS1066** [APDevTeam/Movecraft#582] [Ideas]
As I've mentioned a few times, I think it's time we rewrite Beaming from the ground up (and possibly rename it).  Beaming started as a funny side project of mine to add funny names and an alternative to `/suicide` that players were excessively using to respawn on the Airship Pirates servers.  With the de-dependency work done in Movecraft prior to 7.0, the Essentials integration and respawn mechanics were moved to Beaming, which had already been heavily nerfed due to specific AP functions and developer preferences.

I think there are a few core principles for this rewrite:
1. It should "just work" for most smaller servers out of the box.  This means that:
  - A bed on your piloted ship should work as your respawn point, no crew sign needed <Highest priority>.
  - If you pilot a craft which contains your vanilla spawn point bed, the vanilla spawn point should be set to the new bed location on a normal craft release (Similar to [Beaming/#11](https://github.com/APDevTeam/Beaming/pull/11)) <Second highest priority>.
  - There needs to be some way to designate a "home base" bed, similar to the intent of the priority crewbed.  How exactly you indicate this might be difficult, so I'm open to ideas.  This would be used as your spawn if you either do not have a vanilla spawnpoint, moved your vanilla spawnpoint and sunk, or something else <Second lowest priority>.
  - Only if you do not have any of the above set should the plugin respawn you at the server spawn location <Lowest priority>.
2. Compatibility across versions should be maximized.
3. Code should be well organized (like the Movecraft-Combat rewrite).
4. Multi-crewing should be possible.  I'm not sure how we implement this on beaming, but perhaps this needs API expansions on Movecraft itself.
5. Configure/API all the things.  Obviously there is a point where too much is detrimental, but ideally servers should be able to tailor the plugin to their needs via the config, or at a minimum the API.

I'm not sure if Essentials helps or hinders us here, but at a minimum we should maintain the above principles whether or not we have essentials.

**Comment by goodroach** [comment]
Here's how I think this crew bed system should be done:

1. Any craft that you pilot with a bed and a crew will give you an automatic place to respawn. 
2. Respawn point is temporary and will be your spawn point so any /home or essentials spawn point mechanic will work from there.
3. You have to "opt-in" for this system to work by right-clicking on a crew sign. 
4. OPTIONAL: The number of potential players that can respawn on a ship can be set in the config or by the number of beds
5. The Vanilla spawn point will be your "priority" spawn point instead where you have to right-click on the bed to set your spawn point. 


Not too sure about any potential exploits but this should be the simplest way to go at it.
